### PR TITLE
Fix entropy for uint8

### PR DESCRIPTION
**Pull request by ZhennanQin**
## Description ##
Discussed in https://github.com/apache/incubator-mxnet/pull/13697.

@reminisce @zheng-da  @pengzhao-intel @TaoLv 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by ZhennanQin** [comment]
This is a point fix for uint8 and won't affect GPU accuracy. So test isn't needed.

**Comment by ankkhedia** [comment]
@ZhennanQin Thanks for your contribution!
@mxnet-label-bot add [pr-awaiting-review, Quantization ]

File: tests/python/quantization/test_quantization.py [hunk 1]
```diff
@@ -673,8 +673,9 @@ def test_optimal_threshold_adversarial_case():
     # The worst case for the optimal_threshold function is when the values are concentrated
     # at one edge: [0, 0, ..., 1000]. (histogram)
     # We want to make sure that the optimal threshold in this case is the max.
-    arr = np.array([2]*1000)
-    res = mx.contrib.quant._get_optimal_threshold(arr, num_quantized_bins=5)
+    arr = np.array([2] * 1000)
+    for dtype in ['uint8', 'int8', 'auto']:
+        res = mx.contrib.quant._get_optimal_threshold(arr, dtype, num_quantized_bins=5)
     # The threshold should be 2.
     assert res[3] - 2 < 1e-5
```
Comment (anirudh2290) [inline tests/python/quantization/test_quantization.py]: shouldnt this assert be for each of the dtypes ?

File: python/mxnet/contrib/quantization.py [hunk 2]
```diff
@@ -285,6 +285,10 @@ def _get_optimal_threshold(arr, num_bins=8001, num_quantized_bins=255):
     max_val = np.max(arr)
     th = max(abs(min_val), abs(max_val))
 
+    if min_val >= 0 and quantized_dtype in ['auto', 'uint8']:
```
Comment (anirudh2290) [inline python/mxnet/contrib/quantization.py]: how do the numbers change here for accuracy after this fix: https://cwiki.apache.org/confluence/display/MXNET/MXNet+Graph+Optimization+and+Quantization+based+on+subgraph+and+MKL-DNN

File: tests/python/quantization/test_quantization.py (hunk 1 above)
Comment (ZhennanQin) [reply tests/python/quantization/test_quantization.py]: For uint8, it should have more accuracy, thus the threshold should be half. I will edit here.

File: python/mxnet/contrib/quantization.py (hunk 2 above)
Comment (ZhennanQin) [reply python/mxnet/contrib/quantization.py]: The numbers collected in that link are based on naive calibration method, so they won't be changed after this fix.

File: tests/python/quantization/test_quantization.py (hunk 1 above)
Comment (ZhennanQin) [reply tests/python/quantization/test_quantization.py]: For this case, the ideal result should be `2` no matter what dtype is. So we don't need to use different assert here. @reminisce Do you have other comments here?

**Comment by vandanavk** [comment]
@anirudh2290 is this PR good to go?

File: tests/python/quantization/test_quantization.py (hunk 1 above)
Comment (anirudh2290) [reply tests/python/quantization/test_quantization.py]: can we move the assert inside the for loop in this case ?

**Comment by pengzhao-intel** [comment]
@ZhennanQin Did we resolve all comments in here and is this the valid PR to be merged?

File: tests/python/quantization/test_quantization.py (hunk 1 above)
Comment (ZhennanQin) [reply tests/python/quantization/test_quantization.py]: Done.

**Comment by ZhennanQin** [comment]
@pengzhao-intel @anirudh2290 All comments are resolved now. Thanks for the advice. Can we merge it now?

**Pull request merged**